pkg/server/server.go

```diff
@@ -215,23 +222,47 @@ func (s *Server) handleNar(w http.ResponseWriter, r *http.Request) {
 		narhashStr := chi.URLParam(r, "narhash")
 		narhash, err := nixbase32.DecodeString(narhashStr)
 		if err != nil {
-			http.Error(w, fmt.Sprintf("handle-narinfo: %v", err), http.StatusBadRequest)
+			http.Error(w, fmt.Sprintf("Unable to decode narHash %v: %v", narhashStr, err), http.StatusBadRequest)
 		}
-		r, size, err := s.blobStore.GetBlob(r.Context(), narhash)
+		blobReader, size, err := s.blobStore.GetBlob(r.Context(), narhash)
 		if err != nil {
 			status := http.StatusInternalServerError
 			if err == store.ErrNotFound {
 				status = http.StatusNotFound
 			}
-			http.Error(w, fmt.Sprintf("GET handle-nar: %v", err), status)
+			http.Error(w, fmt.Sprintf("Error retrieving narfile with hash %v: %v", narhashStr, err), status)
 			return
 		}
-		defer r.Close()
+		defer blobReader.Close()
+
+		// check compression suffix, and serve a compressed file depending on that.
+		compressionSuffix := chi.URLParam(r, "compressionSuffix")
+
+		var compressedWriter io.WriteCloser
+		if compressionSuffix != "" {
+			// We only support zstd, gzip, brotli and none, as the others are way too CPU-intensive,
+			// and never advertised anyways.
+			compressedWriter, err = compression.NewCompressorBySuffix(w, compressionSuffix)
+			if err != nil {
+				// We still serve a 404 (as Nix might send a HEAD request while trying to upload xz, for example)
+				http.Error(w, fmt.Sprintf("Unsupported compression suffix: %v", compressionSuffix), http.StatusNotFound)
+			}
+			defer compressedWriter.Close()
+		}
```

Comment:
```suggestion
		var writer io.Writer = w
		if compressionSuffix != "" {
			// We only support zstd, gzip, brotli and none, as the others are way too CPU-intensive,
			// and never advertised anyways.
			compressedWriter, err := compression.NewCompressorBySuffix(w, compressionSuffix)
			if err != nil {
				// We still serve a 404 (as Nix might send a HEAD request while trying to upload xz, for example)
				http.Error(w, fmt.Sprintf("Unsupported compression suffix: %v", compressionSuffix), http.StatusNotFound)
			}
			defer compressedWriter.Close()
			writer = compressedWriter
		}
```
Small nit; by structuring the change like this, the decision-making progress can stay grouped in this place.
(although the code below still has to know to use the `writer` variable).